hub-detect/src/main/groovy/com/blackducksoftware/integration/hub/detect/bomtool/clang/CLangExtractor.java

```diff
@@ -0,0 +1,300 @@
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import com.blackducksoftware.integration.exception.IntegrationException;
+import com.blackducksoftware.integration.hub.bdio.SimpleBdioFactory;
+import com.blackducksoftware.integration.hub.bdio.graph.MutableDependencyGraph;
+import com.blackducksoftware.integration.hub.bdio.model.Forge;
+import com.blackducksoftware.integration.hub.bdio.model.dependency.Dependency;
+import com.blackducksoftware.integration.hub.bdio.model.externalid.ExternalId;
+import com.blackducksoftware.integration.hub.bdio.model.externalid.ExternalIdFactory;
+import com.blackducksoftware.integration.hub.detect.bomtool.ExtractionId;
+import com.blackducksoftware.integration.hub.detect.bomtool.clang.executor.CommandStringExecutor;
+import com.blackducksoftware.integration.hub.detect.bomtool.clang.pkgmgr.PkgMgr;
+import com.blackducksoftware.integration.hub.detect.extraction.model.Extraction;
+import com.blackducksoftware.integration.hub.detect.model.BomToolGroupType;
+import com.blackducksoftware.integration.hub.detect.model.DetectCodeLocation;
+import com.blackducksoftware.integration.hub.detect.util.DetectFileManager;
+import com.blackducksoftware.integration.hub.detect.util.executable.ExecutableRunnerException;
+import com.google.gson.Gson;
+
+@Component
+public class CLangExtractor {
+    private static final String COMPILE_CMD_PATTERN_WITH_DEPENDENCY_OUTPUT_FILE = "%s -M -MF %s";
+    public static final String DEPS_MK_FILENAME_PATTERN = "deps_%s_%d.mk";
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+    private final Set<File> processedDependencyFiles = new HashSet<>(200);
+    private final Set<PackageDetails> processedDependencies = new HashSet<>(40);
+
+    @Autowired
+    private List<PkgMgr> pkgMgrs;
+
+    @Autowired
+    private ExternalIdFactory externalIdFactory;
+
+    @Autowired
+    private CommandStringExecutor executor;
+
+    @Autowired
+    private DependencyFileManager dependencyFileManager;
+
+    @Autowired
+    private DetectFileManager detectFileManager;
+
+    public Extraction extract(final File sourceDir, final ExtractionId extractionId, final File jsonCompilationDatabaseFile) {
+        try {
+            logger.info(String.format("Analyzing %s", jsonCompilationDatabaseFile.getAbsolutePath()));
+            final File outputDirectory = detectFileManager.getOutputDirectory(extractionId);
+            logger.debug(String.format("extract() called; compileCommandsJsonFilePath: %s", jsonCompilationDatabaseFile.getAbsolutePath()));
+            final Set<File> filesForIScan = ConcurrentHashMap.newKeySet(64);
+            final PkgMgr pkgMgr = selectPkgMgr();
```

Comment:
I think this should go in the BomTool and the BomTool can talk to some kind of Manager so you only choose the pkgMgr once even if extraction applies multiple times.